internal/hcsoci/cmd.go

```
@@ -0,0 +1,318 @@
+		// Do not make stdin part of the error group because there is no way for
+		// us or the caller to reliably unblock the c.Stdin read when the
+		// process exits.
+		go func() {
+			_, err := copyAndLog(stdin, c.Stdin, c.Log, "stdin")
+			// Report the stdin copy error. If the process has exited, then the
+			// caller may never see it, but if the error was due to a failure in
+			// stdin read, then it is likely the process is still running.
+			if err != nil {
+				c.stdinErr.Store(err)
+			}
+			// Notify the process that there is no more input.
+			p.CloseStdin()
+		}()
+	}
+
+	if c.Stdout != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stdout, stdout, c.Log, "stdout")
+			return err
+		})
+	}
+
+	if c.Stderr != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stderr, stderr, c.Log, "stderr")
+			return err
+		})
+	}
+
+	if c.Context != nil {
+		go func() {
+			select {
+			case <-c.Context.Done():
+				c.Process.Kill()
+			case <-c.allDoneCh:
+			}
+		}()
+	}
+	return nil
+}
+
+// Wait waits for a command and its IO to complete and closes the underlying
+// process. It can only be called once. It returns an ExitError if the command
+// runs and returns a non-zero exit code.
+func (c *Cmd) Wait() error {
+	err := c.Process.Wait()
+	if err != nil && c.Log != nil {
+		c.Log.WithError(err).Warn("process wait failed")
+	}
+	state := &ExitState{}
+	code, err := c.Process.ExitCode()
+	if err == nil {
+		state.exited = true
+		state.code = code
+	}
+	// Terminate the IO if the copy does not complete in the requested time.
+	if c.CopyAfterExitTimeout != 0 {
+		go func() {
+			t := time.NewTimer(c.CopyAfterExitTimeout)
+			defer t.Stop()
+			select {
+			case <-c.allDoneCh:
+			case <-t.C:
+				// Close the process to cancel any reads to stdout or stderr.
+				c.Process.Close()
+				if c.Log != nil {
+					c.Log.Warn("timed out waiting for stdio relay")
+				}
+			}
+		}()
+	}
+	ioerr := c.iogrp.Wait()
+	if err == nil {
```

This is an odd error order. Can you explain why the cascade works this way?

---

My intention was to follow exec.Cmd's behavior, but I didn't do it right. Their order is:
1. Wait failure
2. Non-zero exit code
3. IO errors

This means that IO errors get swallowed if process execution fails. If we cared to capture these errors, we could make a separate CopyError field that gets filled out... but looking at our code, we don't really care except to log things (which this package will already do).

---

Ok sounds good